lib/Dialect/Seq/Transforms/SinkClockGates.cpp

```diff
@@ -0,0 +1,348 @@
+        else if (auto userGate = dyn_cast<ClockGateOp>(useOp)) {
+          collapseConsecutiveClockGates(clkGate, userGate);
+        } else {
+          opsWithGatedClock.insert(useOp);
+        }
+      }
+    });
+  };
+  // Seed phase, find the instances to start the traversal from.
+  // These instances have a gated clock as input.
+  for (auto *node : *graph) {
+    if (auto hwModule = node->getModule<HWModuleOp>())
+      findInstancesWithGatedClock(hwModule);
+  }
+  // Get the initial set of clock gates which this pass tries to sink.
+  numClockGatesConverted = instancesWithGatedClk.size();
+
+  // Transformation phase, push the clock_gate through the instances and find
+  // all the indirect instances that have a gated clock.
+
+  // Iterate until none of the instances have a gated clock as input.
+  while (!instancesWithGatedClk.empty()) {
+
+    // Remove an entry from the set.
+    const auto *entry = instancesWithGatedClk.begin();
+    HWInstanceLike inst = *entry;
+    instancesWithGatedClk.erase(entry);
+
+    // Now, find all the gated clocks that are input to the instance, and
+    // duplicate it at the referenced module.
+
+    SmallVector<HWInstanceLike, 2> moduleInstances;
+    HWModuleOp refMod = getReferencedModules(inst, moduleInstances);
+    if (!refMod) {
+      // If we cannot identify a unique HWModuleOp, ignore the instance.
+      // Multiple references and extern modules cannot be handled, and gated
+      // clocks cannot be sunk into them.
+      continue;
+    }
+    // Index of the instance operands that are a gated clock.
+    SmallVector<unsigned> gatedClkPorts;
+    // The new enable ports that will be added to the instance.
+    SmallVector<std::pair<StringAttr, Value>> enablePorts;
+    // The instance may or maynot have argNames attribute.
+    SmallVector<StringAttr> argNames = getInstanceArgNames(inst);
+    unsigned enablePortNum = inst->getNumOperands();
+    ClockEnablePortPairs &clockEnablePortIndices =
+        moduleClockGateEnableMap[refMod];
+    // Iterate over all instance operands to find a gated clock.
+    for (auto [index, in] : llvm::enumerate(inst->getOperands())) {
+      if (!isa<seq::ClockType>(in.getType()))
+        continue;
+      auto clkGate = dyn_cast_or_null<ClockGateOp>(in.getDefiningOp());
+      if (!clkGate)
+        continue;
+      if (clkGate->hasOneUse())
+        opsToErase.insert(clkGate);
+      // Replace the gated clock with the original base clock.
+      inst->setOperand(index, clkGate.getInput());
+      Value enable = getEnable(clkGate);
+      if (clockEnablePortIndices.contains(index)) {
+        // The enable for this clock is already added, (from another instance of
+        // the module).
+        auto oldEnablePort = clockEnablePortIndices[index];
+        auto oldEnableVal = inst->getOperand(oldEnablePort);
+        auto *def = oldEnableVal.getDefiningOp();
+        assert(llvm::isa_and_nonnull<ConstantOp>(def));
```

Comment:
Why is it guaranteed to be constant?

Reply:
This is a constant, because this is another instance of a module that is already visited. When the first instance was visited, we added the enable port to all the module instances. The enable port for all the other instances were set to a constant true, except the one that is being visited, which is set to the clock gate op's enable. This ensures that we donot add multiple enable ports for the same clock gate port when visiting multiple instances of a module.